### PR TITLE
pnpm: 9.12.1 -> 9.12.2

### PR DESCRIPTION
**Pull request by gepbird**
Release: https://github.com/pnpm/pnpm/releases/tag/v9.12.2

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by gepbird** [comment]
Most packages fail with ` ERR_PNPM_NO_OFFLINE_TARBALL  A package is missing from the store but cannot download it in offline mode. The missing package may be downloaded from https://registry.npmjs.org/xxx/-/xxx.tgz.`
I'll investigate this on the weekend when I'll have time.

Result of `nixpkgs-review pr 349093` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>42 packages failed to build:</summary>
  <ul>
    <li>astro-language-server</li>
    <li>authelia</li>
    <li>autoprefixer</li>
    <li>cdxgen</li>
    <li>clash-verge-rev</li>
    <li>en-croissant</li>
    <li>follow</li>
    <li>gitbutler</li>
    <li>goofcord</li>
    <li>heroic</li>
    <li>heroic-unwrapped</li>
    <li>homebox</li>
    <li>legcord</li>
    <li>lemmy-ui</li>
    <li>metacubexd</li>
    <li>misskey</li>
    <li>n8n</li>
    <li>overlayed</li>
    <li>pgrok</li>
    <li>pgrok.server</li>
    <li>postcss-cli</li>
    <li>renovate</li>
    <li>rsshub</li>
    <li>siyuan</li>
    <li>sketchybar-app-font</li>
    <li>stylelint-lsp</li>
    <li>surrealist</li>
    <li>syncyomi</li>
    <li>taler-challenger</li>
    <li>taler-merchant</li>
    <li>taler-sync</li>
    <li>taler-wallet-core</li>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
    <li>vencord</li>
    <li>vencord-web-extension</li>
    <li>vesktop</li>
    <li>vikunja</li>
    <li>webcord-vencord</li>
    <li>wrangler</li>
    <li>youtube-music</li>
    <li>zenn-cli</li>
  </ul>
</details>
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>pnpm</li>
    <li>pot</li>
  </ul>
</details>

**Comment by NyCodeGHG** [comment]
Looks like upstream changed their store format, which broke most of our FOD hashes :/

**Comment by gepbird** [comment]
> Looks like upstream changed their store format, which broke most of our FOD hashes :/

Running a diff for example on the old and new `astro-language-server.pnpmDeps`, shows only a file name change, file content is the same:
```
Only in /nix/store/mbawgip8msszvjkqhsjnhv0bwxj89cqv-astro-language-server-pnpm-deps/v3/files/8a: 4e547ef22ddf054acadfe74ec7e8a083af03d7fdfe918fcf0b41d155f0e3163c58f94097e4fffeffa628ded2de364f9620d0208bbe7e5ac9c01073c374aa26
Only in /nix/store/kn62rkbq14mkah57sadr3kzc58hn8dm4-astro-language-server-pnpm-deps/v3/files/8a: 4e547ef22ddf054acadfe74ec7e8a083af03d7fdfe918fcf0b41d155f0e3163c58f94097e4fffeffa628ded2de364f9620d0208bbe7e5ac9c01073c374aa26-exec
```

I'll update all the pnpm hashes soon.

**Comment by gepbird** [comment]
`teleport` failure is unrelated: https://hydra.nixos.org/build/274984883

## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 349093`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 42 packages built:</summary>
  <ul>
    <li>astro-language-server</li>
    <li>authelia</li>
    <li>autoprefixer</li>
    <li>cdxgen</li>
    <li>clash-verge-rev</li>
    <li>en-croissant</li>
    <li>follow</li>
    <li>gitbutler</li>
    <li>goofcord</li>
    <li>heroic</li>
    <li>heroic-unwrapped</li>
    <li>homebox</li>
    <li>legcord</li>
    <li>lemmy-ui</li>
    <li>metacubexd</li>
    <li>misskey</li>
    <li>n8n</li>
    <li>overlayed</li>
    <li>pgrok</li>
    <li>pgrok.server</li>
    <li>pnpm</li>
    <li>postcss-cli</li>
    <li>pot</li>
    <li>renovate</li>
    <li>rsshub</li>
    <li>siyuan</li>
    <li>sketchybar-app-font</li>
    <li>stylelint-lsp</li>
    <li>surrealist</li>
    <li>syncyomi</li>
    <li>taler-challenger</li>
    <li>taler-merchant</li>
    <li>taler-sync</li>
    <li>taler-wallet-core</li>
    <li>vencord</li>
    <li>vencord-web-extension</li>
    <li>vesktop</li>
    <li>vikunja</li>
    <li>webcord-vencord</li>
    <li>wrangler</li>
    <li>youtube-music</li>
    <li>zenn-cli</li>
  </ul>
</details>

File: pkgs/development/tools/pnpm/default.nix
Comment (Scrumplex) [inline pkgs/development/tools/pnpm/default.nix:1]: Could you add yourself as a maintainer to the derivation?

Also, we can use the new OWNERS file to get pinged for changes in `pkgs/development/tools/pnpm` so it would be great if you can add yourself and @Scrumplex to that file ^^

**Comment by Scrumplex** [comment]
podman-desktop from #343648 needs to be updated

**Comment by Scrumplex** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 349093`

---
### `x86_64-linux`
<details>
  <summary>:x: 3 packages failed to build:</summary>
  <ul>
    <li>podman-desktop</li>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 42 packages built:</summary>
  <ul>
    <li>astro-language-server</li>
    <li>authelia</li>
    <li>autoprefixer</li>
    <li>cdxgen</li>
    <li>clash-verge-rev</li>
    <li>en-croissant</li>
    <li>follow</li>
    <li>gitbutler</li>
    <li>goofcord</li>
    <li>heroic</li>
    <li>heroic-unwrapped</li>
    <li>homebox</li>
    <li>legcord</li>
    <li>lemmy-ui</li>
    <li>metacubexd</li>
    <li>misskey</li>
    <li>n8n</li>
    <li>overlayed</li>
    <li>pgrok</li>
    <li>pgrok.server</li>
    <li>pnpm (pnpm_9)</li>
    <li>postcss-cli</li>
    <li>pot</li>
    <li>renovate</li>
    <li>rsshub</li>
    <li>siyuan</li>
    <li>sketchybar-app-font</li>
    <li>stylelint-lsp</li>
    <li>surrealist</li>
    <li>syncyomi</li>
    <li>taler-challenger</li>
    <li>taler-merchant</li>
    <li>taler-sync</li>
    <li>taler-wallet-core</li>
    <li>vencord</li>
    <li>vencord-web-extension</li>
    <li>vesktop</li>
    <li>vikunja</li>
    <li>webcord-vencord</li>
    <li>wrangler</li>
    <li>youtube-music</li>
    <li>zenn-cli</li>
  </ul>
</details>

**Comment by gepbird** [comment]
Updated podman-desktop and metacubexd hashes

**Comment by Scrumplex** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 349093`

---
### `x86_64-linux`
<details>
  <summary>:x: 4 packages failed to build:</summary>
  <ul>
    <li>metacubexd</li>
    <li>podman-desktop</li>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 41 packages built:</summary>
  <ul>
    <li>astro-language-server</li>
    <li>authelia</li>
    <li>autoprefixer</li>
    <li>cdxgen</li>
    <li>clash-verge-rev</li>
    <li>en-croissant</li>
    <li>follow</li>
    <li>gitbutler</li>
    <li>goofcord</li>
    <li>heroic</li>
    <li>heroic-unwrapped</li>
    <li>homebox</li>
    <li>legcord</li>
    <li>lemmy-ui</li>
    <li>misskey</li>
    <li>n8n</li>
    <li>overlayed</li>
    <li>pgrok</li>
    <li>pgrok.server</li>
    <li>pnpm</li>
    <li>postcss-cli</li>
    <li>pot</li>
    <li>renovate</li>
    <li>rsshub</li>
    <li>siyuan</li>
    <li>sketchybar-app-font</li>
    <li>stylelint-lsp</li>
    <li>surrealist</li>
    <li>syncyomi</li>
    <li>taler-challenger</li>
    <li>taler-merchant</li>
    <li>taler-sync</li>
    <li>taler-wallet-core</li>
    <li>vencord</li>
    <li>vencord-web-extension</li>
    <li>vesktop</li>
    <li>vikunja</li>
    <li>webcord-vencord</li>
    <li>wrangler</li>
    <li>youtube-music</li>
    <li>zenn-cli</li>
  </ul>
</details>

**Comment by gepbird** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 349093`

---
### `x86_64-linux`
<details>
  <summary>:x: 2 packages failed to build:</summary>
  <ul>
    <li>teleport (teleport_16)</li>
    <li>teleport.client (teleport_16.client)</li>
  </ul>
</details>
<details>
  <summary>:white_check_mark: 43 packages built:</summary>
  <ul>
    <li>astro-language-server</li>
    <li>authelia</li>
    <li>autoprefixer</li>
    <li>cdxgen</li>
    <li>clash-verge-rev</li>
    <li>en-croissant</li>
    <li>follow</li>
    <li>gitbutler</li>
    <li>goofcord</li>
    <li>heroic</li>
    <li>heroic-unwrapped</li>
    <li>homebox</li>
    <li>legcord</li>
    <li>lemmy-ui</li>
    <li>metacubexd</li>
    <li>misskey</li>
    <li>n8n</li>
    <li>overlayed</li>
    <li>pgrok</li>
    <li>pgrok.server</li>
    <li>pnpm</li>
    <li>podman-desktop</li>
    <li>postcss-cli</li>
    <li>pot</li>
    <li>renovate</li>
    <li>rsshub</li>
    <li>siyuan</li>
    <li>sketchybar-app-font</li>
    <li>stylelint-lsp</li>
    <li>surrealist</li>
    <li>syncyomi</li>
    <li>taler-challenger</li>
    <li>taler-merchant</li>
    <li>taler-sync</li>
    <li>taler-wallet-core</li>
    <li>vencord</li>
    <li>vencord-web-extension</li>
    <li>vesktop</li>
    <li>vikunja</li>
    <li>webcord-vencord</li>
    <li>wrangler</li>
    <li>youtube-music</li>
    <li>zenn-cli</li>
  </ul>
</details>

**Review by Scrumplex — APPROVED** [review]
LGTM. We should notify open PRs about this change so that they can update their hashes.

**Comment by gepbird** [comment]
> LGTM. We should notify open PRs about this change so that they can update their hashes.

I'll do that after this is merged, also they should get a merge conflict for the hashes too.

**Comment by gepbird** [comment]
Thanks for notifying them about the hash changes!

I also checked which PR got broken by this and found the same ones as you did, except for `dorion`.

**Comment by Aleksanaa** [comment]
Is this hash change unavoidable? This type of behavior should be kept to a minimum.

**Comment by dotlambda** [comment]
> Looks like upstream changed their store format, which broke most of our FOD hashes :/

This is why `pnpm.fetchDeps` should never have been merged. Imagine thousands of packages using it and a change like this quickly becomes unsustainable.
Somebody urgently needs to write a different FOD fetcher, simiilar to the one for npm, whose hash doesn't change upon version bumps.

**Comment by dotlambda** [comment]
Alternatively, can we ask upstream to commit to not break the hashes outside of major version bumps? And write a CI test for upstream that ensures this?

**Comment by Aleksanaa** [comment]
> Running a diff for example on the old and new `astro-language-server.pnpmDeps`, shows only a file name change, file content is the same:
> 
> ```
> Only in /nix/store/mbawgip8msszvjkqhsjnhv0bwxj89cqv-astro-language-server-pnpm-deps/v3/files/8a: 4e547ef22ddf054acadfe74ec7e8a083af03d7fdfe918fcf0b41d155f0e3163c58f94097e4fffeffa628ded2de364f9620d0208bbe7e5ac9c01073c374aa26
> Only in /nix/store/kn62rkbq14mkah57sadr3kzc58hn8dm4-astro-language-server-pnpm-deps/v3/files/8a: 4e547ef22ddf054acadfe74ec7e8a083af03d7fdfe918fcf0b41d155f0e3163c58f94097e4fffeffa628ded2de364f9620d0208bbe7e5ac9c01073c374aa26-exec
> ```


But is it possible for us to at least "fixup" them to something simpler, so it doesn't matter?

Or is this completely happening in upstream servers and we don't have any control of it?

**Comment by Scrumplex** [comment]
> But is it possible for us to at least "fixup" them to something simpler, so it doesn't matter?

If we had renamed these files back to what pnpm < 9.12.2 used, the configure hook would fail installing dependencies as pnpm 9.12.2 wouldn't find the expected files in its store.

Maybe we could have fixed them up and then unfixed them during configuration. But I think that wouldn't be very maintainable in the long run.

> Alternatively, can we ask upstream to commit to not breaking the hashes outside of major version bumps? And write a CI test for upstream that ensures this?

I think this is a good way to go. If upstream agrees that a change to the store is a breaking change, they might just bump to a new major version every time the store format changes. In those cases, we can just release a new `pnpm_<major>` package and update all references in Nixpkgs to the previous version until packages update.

**Comment by Scrumplex** [comment]
I created an issue upstream to discuss this: https://github.com/pnpm/pnpm/issues/8713

**Comment by Aleksanaa** [comment]
One more thing: we may need a release note for this, as it's somehow breaking.

**Comment by gepbird** [comment]
> This is why pnpm.fetchDeps should never have been merged. Imagine thousands of packages using it and a change like this quickly becomes unsustainable.
Somebody urgently needs to write a different FOD fetcher, simiilar to the one for npm, whose hash doesn't change upon version bumps.

Until we don't have that many pnpm packages (currently there are about 50), mass hash changes  are not the end of the world if it only happens 1 or 2 times in a year.

Package update PRs that use `pnpm.fetchDeps` get a merge conflict, so most likely the PR author will regenerate the hash. From what I've seen it didn't cause that big of an issue and was dealt with my active PR authors.

Hovewer package init PRs won't get merge conflicts and we need to hunt these down and notify the author, this is where we could've done better. Searching for `pnpm.fetchDeps` or even `pnpm` in nixpkgs/pulls is not sufficient. With a tool that could properly search in PR diffs, this could've been mitigated (at least in nixpkgs and not in nur or other projects relying on this). Do we have anything like that for searching PR diffs?


> Alternatively, can we ask upstream to commit to not breaking the hashes outside of major version bumps? And write a CI test for upstream that ensures this?

Exactly, this should've been only released in the next major version.

> One more thing: we may need a release note for this, as it's somehow breaking.

Mostly for people using `pnpm.fetchDeps` outside of nixpkgs, this will be very helpful. Can someone make a PR? I won't have the time for that in the near future.

**Pull request merged**
